internal/generator/kzg/template/marshal.go.tmpl

```diff
@@ -72,6 +75,88 @@ func (vk *VerifyingKey) writeTo(w io.Writer, options ...func(*{{.CurvePackage}}.
 	return enc.BytesWritten(), nil
 }
 
+// UnsafeToBytes returns the binary encoding of the entire SRS memory representation
+// It is meant to be use to achieve fast serialization/deserialization and
+// is not compatible with WriteTo / ReadFrom. It does not do any validation
+// and doesn't encode points in a canonical form.
+// @unstable: the format may change in the future
+// If maxPkPoints is provided, the number of points in the ProvingKey will be limited to maxPkPoints
+func (srs *SRS) UnsafeToBytes(maxPkPoints ...int) ([]byte, error) {
+	maxG1 := len(srs.Pk.G1)
+	if len(maxPkPoints) > 0 && maxPkPoints[0] < maxG1 && maxPkPoints[0] > 0{
+		maxG1 = maxPkPoints[0]
+	}
+	// first we write the VerifyingKey; it is small so we re-use WriteTo
+	var buf bytes.Buffer
+
+	if _, err := srs.Vk.writeTo(&buf, {{.CurvePackage}}.RawEncoding()); err != nil {
+		return nil, err
+	}
+
+	buf.Grow(2 * maxG1 * fp.Bytes + 8) // pre-allocate space for the ProvingKey
+
+	// write nb points we encode.
+	if err := binary.Write(&buf, binary.LittleEndian, uint64(maxG1)); err != nil {
+		return nil, err
+	}
+
+	// write the limbs directly
+	var bbuf [fp.Bytes * 2]byte
+	for i := 0; i < maxG1; i++ {
+		for j:= 0 ; j < fp.Limbs; j++ {
+			binary.LittleEndian.PutUint64(bbuf[j*8:j*8+8], srs.Pk.G1[i].X[j])
+		}
+		for j:= 0 ; j < fp.Limbs; j++ {
+			binary.LittleEndian.PutUint64(bbuf[fp.Bytes + j*8:fp.Bytes + j*8+8], srs.Pk.G1[i].Y[j])
+		}
+		if _, err := buf.Write(bbuf[:]); err != nil {
+			return nil, err
+		}
+	}
+	return buf.Bytes(), nil
+}
+
+// UnsafeFromBytes deserializes the SRS from a byte slice
+// It is meant to be use to achieve fast serialization/deserialization and
+// is not compatible with WriteTo / ReadFrom. It does not do any validation
+// and doesn't encode points in a canonical form.
+// @unstable: the format may change in the future
+func (srs *SRS) UnsafeFromBytes(data []byte, maxPkPoints ...int) error {
+	buf := bytes.NewReader(data)
+
+	// first we read the VerifyingKey; it is small so we re-use ReadFrom
+	if _, err := srs.Vk.ReadFrom(buf); err != nil {
+		return err
+	}
+
+	// read nb points we encode.
+	var nbPoints uint64
+	if err := binary.Read(buf, binary.LittleEndian, &nbPoints); err != nil {
+		return err
+	}
+
+	if len(maxPkPoints) == 1 && maxPkPoints[0] > 0 && int(nbPoints) > maxPkPoints[0] {
+		nbPoints = uint64(maxPkPoints[0])
+	}
+
+	srs.Pk.G1 = make([]{{.CurvePackage}}.G1Affine, nbPoints)
+
+	// read the limbs directly
+	var bbuf [fp.Bytes * 2]byte
+	for i := 0; i < int(nbPoints); i++ {
+		if _, err := io.ReadFull(buf, bbuf[:]); err != nil {
```

Comment:
`srs.Vk.ReadFrom` returns number of bytes read and it is 8 byte for the length. Instead of `io.ReadFull` we can just set `bbuf = buf[n+8+i*2*fp.Bytes:n+8+(i+1)*2*fp.Bytes]`